main/src/main/scala/sbt/internal/ClassLoaders.scala

```diff
@@ -156,10 +156,25 @@ private[sbt] object ClassLoaders {
         val cpFiles = fullCP.map(_._1)
 
         val allDependencies = cpFiles.filter(allDependenciesSet)
-        val scalaReflectJar = allDependencies.collectFirst {
-          case f if f.getName == "scala-reflect.jar" =>
-            si.allJars.find(_.getName == "scala-reflect.jar")
-        }.flatten
+        /*
+         * We need to put scalatest and its dependencies (scala-xml) in its own layer so that the
+         * classloader for scalatest is generally not ever closed. This is because one project
+         * closing its scalatest loader can cause a crash in another project if that project
+         * is reading the scalatest bundle properties because the underlying JarFile instance is
+         * shared across URLClassLoaders if it points to the same jar.
+         *
+         * TODO: Make this configurable and/or add other frameworks in this layer.
+         */
+        val (frameworkDependencies, otherDependencies) = allDependencies.partition { d =>
+          val name = d.getName
+          name.startsWith("scalatest_") || name.startsWith("scalactic_") ||
+          name.startsWith("scala-xml_")
+        }
```

Comment:
What makes ScalaTest so special? Could this potentially happen to anything that reads from a resource input stream?

Reply:
Theoretically this issue could come up with any jar that is shared across multiple projects. I also don't know if it affects classloading or only resource loading. I will also make a separate PR that re-enables `-Dsbt.classloader.close`. At the moment it doesn't do anything because I mistakenly thought that zombie classloaders would solve the issue. If closing a classloader in one project can cause an independent classloader to fail to load classes in another project, users should be able to disable closing the loaders. Given that this is the only reported manifestation of the problem, I'm not convinced that its widespread enough to disable closing the classloaders by default. It's possible though that is the route we should ultimately take. As I mentioned in an issue comment, the other approach would be to copy the classpath ala `run` in `test`. There is merit to this idea but I worry about the io and disk space overhead.

Scalatest is special because of its ubiquity in the scala ecosystem. I could see an argument for doing this for all of the default testFrameworks: ScalaCheck, Specs, Specs2 and JUnit. It's just kind of a pain because you have to include all of the framework dependencies in this layer. Because the testFrameworks are specified by class name, we don't know a priori what library dependency provides the implementation which makes auto identifying the necessary jars difficult.

Other notes:
I could only reproduce the issue with jdk8 so it isn't clear to me how likely closing the loaders is to cause problems in newer jdks. If it isn't an issue in newer jdks, we could consider only closing the classloaders given some minimum jdk threshold.

Independent of the resource loading issue, this also brings a performance improvement to builds that use scalatest without turbo. In my experience, scalatest is by far the heaviest test framework so the performance benefit would be more modest with say utest or minitest.

Reply:
After writing this, it occurred to me that another option would be to enqueue classloaders for closing after test completes rather than directly close them. Then we could actually perform the closing when we enter the shell.